src/main/java/org/spongepowered/api/event/item/inventory/EnchantItemEvent.java

```diff
@@ -0,0 +1,114 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.event.item.inventory;
+
+import org.spongepowered.api.data.DataSerializable;
+import org.spongepowered.api.data.Transaction;
+import org.spongepowered.api.item.enchantment.Enchantment;
+import org.spongepowered.api.item.inventory.ItemStack;
+import org.spongepowered.api.item.inventory.Slot;
+
+import java.util.List;
+
+/**
+ * This event is triggered when an item is enchanted at an enchanting table.
+ */
+public interface EnchantItemEvent extends ClickInventoryEvent {
+
+    /**
+     * Returns the slot of the enchanted item.
+     *
+     * @return the slot if the enchanted item.
+     */
+    Slot getSlot();
+
+    /**
+     * Triggers when the enchantment costs for an item are calculated.
+     */
+    // Mixin onCraftMatrixChanged in loop after EnchantmentHelper.calcItemStackEnchantability
+    // Forge: ForgeEventFactory.onEnchantmentLevelSet
+    interface CalculateCost extends TargetContainerEvent {
+
+        int getSeed(); // pseudo random seed
+        int getSlot(); // 1,2,3
+        int getPower(); // based on books (0-?)
+
+        ItemStack getItem(); // item to enchant
+
+        // TODO Transaction that allows non DataSerializable types?
+        Transaction<EnchantmentCost> getCost();
+
+        interface EnchantmentCost extends DataSerializable {
+            int cost();
+            // vanilla max 30
+            // going over 30 a lot may break vanilla enchanting
+            EnchantmentCost withCost(int cost);
+        }
+
+    }
+
+    /**
+     * Triggers when the enchantments for an item are calculated.
+     * This happens after a new item to enchant is put in the table
+     * and again when the item is enchanted.
+     */
+    // Mixin ContaienrEnchantment.getEnchantmentList and/or
+    //       EnchantmentHelper.buildEnchantmentList-at-getEnchantmentDatas
+    interface CalculateEnchantment extends TargetContainerEvent {
+
+        // Seed + slot are seed for enchantment-list generation
+        int getSeed();
+        // 1,2,3
+        int getSlot();
+        // vanilla max 30
+        int getCost();
+
+        // First Enchantment in list is used for clue in client view
+        Transaction<EnchantmentList> getEnchantments();
+
+        interface EnchantmentList extends DataSerializable {
+
+            // Utility functions to generate different enchantment lists
+
+            List<Enchantment> enchantments();
+            // fixed enchantments defined by plugin
+            EnchantmentList withEnchantments(Enchantment... enchantments);
+            EnchantmentList withEnchantments(List<Enchantment> enchantments);
```

Comment:
Why a List ?

Reply:
You add multiple enchantments in one go?

Reply:
uh, I commented on the wrong line.

Anyway, why not a Collection ?

Can you have duplicate enchantments ?

Reply:
The first enchantment of the list is taken for the preview.
Duplicate enchantements probably makes no sense.

Reply:
IMO, that's an implementation details.

Reply:
I feel a `List` is still better since it imperatively retains order, whereas a `Collection` is a hit and miss of ordering. The aspect of "first enchantment is used for preview" is kinda meh, but it would make sense if we keep a `List` and just state something about it being used as a preview because plugin developers otherwise may not know that fact beforehand.

Reply:
We can just use an `Iterable` and document our stand on the iteration order.